### PR TITLE
add new template helm-controller and source-controller

### PR DESCRIPTION
**Pull request by HansCeril**
add makefile
forgot makefile
add templated variables
add roles clusterroles rolebinding clusterrolebinding serviceaccount
correction

### CODE REVIEW

File: .vscode/configurationCache.log [hunk 1]
```diff
@@ -0,0 +1 @@
+{"buildTargets":[],"launchTargets":[],"customConfigurationProvider":{"workspaceBrowse":{"browsePath":[],"compilerArgs":[]},"fileIndex":[]}}
```
Comment (swbsf) [inline .vscode/configurationCache.log:1]: .vscode folder should be in .gitignore

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (swbsf) [inline .vscode/settings.json:1]: .vscode folder should be in .gitignore

File: .vscode/targets.log [hunk 3]
```diff
@@ -0,0 +1,231 @@
+make: *** No rule to make target 'all'.  Stop.
```
Comment (swbsf) [inline .vscode/targets.log:1]: .vscode folder should be in .gitignore

File: charts/helm-controller-operator/values.yaml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      maxUnavailable: 25%
+    type: RollingUpdate
+  sourceControllerStrategy: Recreate
+
+
+service:
+  type: ClusterIP
+  port: 80
+  clusterIps:
+    - 10.96.163.142
```
Comment (swbsf) [inline charts/helm-controller-operator/values.yaml:79]: Maybe set your IDE to add last line on save?

File: charts/helm-controller-operator/templates/roles.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  resources:
+  - leases
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - patch
+  - delete
```
Comment (swbsf) [inline charts/helm-controller-operator/templates/roles.yaml:44]: Maybe set your IDE to add last line?

File: charts/helm-controller-operator/Makefile [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ifeq (, $(shell which yq))
+	@pip3 install yq
+endif
+YQ=$(shell which yq)
+
+# find or download kustomize
+# download kustomize if necessary
+kustomize:
+ifeq (, $(shell which kustomize))
+	@curl -s "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh" | bash -s -- "4.4.1" "$$(go env GOPATH)/bin"
```
Comment (swbsf) [inline charts/helm-controller-operator/Makefile:48]: Maybe move the version to a global variable ?

File: charts/helm-controller-operator/Chart.yaml [hunk 7]
```diff
@@ -0,0 +1,24 @@
+apiVersion: v2
+name: helm-controller-operator
+description: A Helm chart for Kubernetes
```
Comment (vincentmrg) [inline charts/helm-controller-operator/Chart.yaml:3]: A better description would be nice!

File: charts/helm-controller-operator/Makefile [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+endif
+KUSTOMIZE=$(shell which kustomize)
```
Comment (vincentmrg) [inline charts/helm-controller-operator/Makefile:50]:
```suggestion
KUSTOMIZE=$(shell which kustomize)

```

File: charts/helm-controller-operator/templates/cluster_roles_bidings.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: helm-manager-rolebinding
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/cluster_roles_bidings.yaml:5]:
```suggestion
  name: {{ include "helm-manager.fullname" . }}-rolebinding
```

File: charts/helm-controller-operator/templates/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+    matchLabels:
+      app: helm-controller
+  {{- with .Values.manager }}
+  strategy:
+    {{- toYaml .helmControllerStrategy | nindent 4 }}
+  template:
+    metadata:
+      annotations:
+        prometheus.io/port: "8080"
+        prometheus.io/scrape: "true"
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/deployment.yaml:21]: Should not be hardcoded
```suggestion
      {{- with .Values.podAnnotations }}
      annotations:
        {{- toYaml . | nindent 8 }}
      {{- end }}
```

File: charts/helm-controller-operator/templates/deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+      labels:
+        app: helm-controller
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/deployment.yaml:23]: Cf : use the function from the tpl generated by helm
```suggestion
      labels:
        {{- include "helm-controller-operator.selectorLabels" . | nindent 8 }}
```

File: charts/helm-controller-operator/templates/deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        - --log-encoding=json
+        - --enable-leader-election
+        env:
+        - name: RUNTIME_NAMESPACE
+          valueFrom:
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+        {{- with .Values.manager.imageHelmController.image }}
+        image: "{{ .repository }}:{{ .tag }}"
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/deployment.yaml:41]:
```suggestion
        image: "{{ .repository }}:{{ {{ .tag | default $.Chart.AppVersion }} }}"
```

File: charts/helm-controller-operator/templates/deployment.yaml [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+          allowPrivilegeEscalation: false
+          capabilities:
+            drop:
+            - ALL
+          readOnlyRootFilesystem: true
+          runAsNonRoot: true
+          seccompProfile:
+            type: RuntimeDefault
+        terminationMessagePath: /dev/termination-log
+        terminationMessagePolicy: File
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/deployment.yaml:70]:
```suggestion
```

File: charts/helm-controller-operator/templates/deployment.yaml [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+        volumeMounts:
+        - mountPath: /tmp
+          name: temp
+      dnsPolicy: ClusterFirst
+      restartPolicy: Always
+      schedulerName: default-scheduler
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/deployment.yaml:76]:
```suggestion
```

File: charts/helm-controller-operator/templates/deployment.yaml [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+      securityContext:
+        fsGroup: 1337
+      terminationGracePeriodSeconds: 600
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/deployment.yaml:79]:
```suggestion
      terminationGracePeriodSeconds: {{ .Values.terminationGracePeriod }}
```

File: charts/helm-controller-operator/templates/helmController/service.yaml [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+      targetPort: https-metrics
+      protocol: TCP
+    {{- else }}
+    - name: http-metrics
+      port: 8080
+      targetPort: http-metrics
+      protocol: TCP
+    {{- end }}
+  selector:
+    {{- include "helm-controller-operator.selectorLabels" . | nindent 4 }}
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/helmController/service.yaml:22]: When you describe this service do you see all its pods in the status or pods from the sourceController as well ?
Hmm I think you should have different selector labels for the two deployment, e.g. by adding `app.kubernetes.io/component: helm-controller` to labels and selectors (e.g. as the [chart we saw does](https://github.com/grafana/helm-charts/blob/main/charts/tempo-distributed/templates/querier/_helpers-querier.tpl))

File: charts/helm-controller-operator/values.yaml [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+  # -- Override rtmp operator ServiceAccount name.
+  name: ""
+  # -- Annotations added to the ServiceMonitor.
+  annotations: {}
+
+service:
+  type: ClusterIP
+  port: 80
+  clusterIps:
+    - 10.96.163.142
```
Comment (antonincms) [inline charts/helm-controller-operator/values.yaml:175]:
```suggestion
```

File: charts/helm-controller-operator/templates/sourceController/service.yaml [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+spec:
+  clusterIP: {{ (index .Values.service.clusterIps 0) }}
+  clusterIPs:
+  {{- range .Values.service.clusterIps }}
+    - {{ . }}
+  {{- end }}
+  internalTrafficPolicy: Cluster
+  ipFamilies:
+  - IPv4
+  ipFamilyPolicy: SingleStack
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/sourceController/service.yaml:41]: All of this is not needed : 
* clusterIP are automatically filled
* internalTrafficPolicy / ipFamilies / ipFamilyPolicy are at their default values. 

You should compare what helm template and the kustomize produces, I'm sure theses were not there.

```suggestion
```

File: charts/helm-controller-operator/templates/sourceController/service.yaml [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+  ports:
+  - name: http
+    port: 80
+    protocol: TCP
+    targetPort: http
+  selector:
+    {{- include "helm-controller-operator.selectorLabels" . | nindent 4 }}
+  sessionAffinity: None
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/sourceController/service.yaml:49]:
```suggestion
```

File: charts/helm-controller-operator/values.yaml [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+  manager:
+    # -- Number of desired pods.
+    replicaCount: 1
+    # -- Pods annotation.
+    podannotations: {}
+      # prometheus.io/port: "8080"
+      # prometheus.io/scrape: "true"
+    terminationGracePeriodSeconds: 600
+    progressDeadlineSeconds: 600
+    revisionHistoryLimit: 10
```
Comment (antonincms) [inline charts/helm-controller-operator/values.yaml:17]: What are these ? I don't see where they are used.

File: charts/helm-controller-operator/values.yaml [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+        path: /readyz
+        port: healthz
+        scheme: HTTP
+    ressources:
+      limits:
+        cpu: "1"
+        memory: 1Gi
+      requests:
+        cpu: 100m
+        memory: 64Mi
```
Comment (antonincms) [inline charts/helm-controller-operator/values.yaml:49]: No resources by default (and resources takes one s)
```suggestion
    resources: {}
    #  limits:
    #    cpu: "1"
    #    memory: 1Gi
    #  requests:
    #    cpu: 100m
    #    memory: 64Mi
```

File: charts/helm-controller-operator/templates/helmController/deployment.yaml [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+        - containerPort: 9440
+          name: healthz
+          protocol: TCP
+        {{- with .Values.helmController.manager }}
+        livenessProbe:
+          {{- toYaml .livenessProbe | nindent 10 }}
+        readinessProbe:
+          {{- toYaml .readinessProbe | nindent 10 }}
+        resources:
+          {{- toYaml .ressources | nindent 10 }}
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/helmController/deployment.yaml:50]:
```suggestion
          {{- toYaml .resources | nindent 10 }}
```

File: charts/helm-controller-operator/values.yaml [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+  manager:
+    # -- Number of desired pods.
+    replicaCount: 1
+    # -- Pods annotation.
+    podannotations: {}
+      # prometheus.io/port: "8080"
+      # prometheus.io/scrape: "true"
+    terminationGracePeriodSeconds: 10
+    progressDeadlineSeconds: 600
+    revisionHistoryLimit: 10
```
Comment (antonincms) [inline charts/helm-controller-operator/values.yaml:78]: Idem, what is this

File: charts/helm-controller-operator/values.yaml [hunk 24]
```diff
@@ -0,0 +1,175 @@
...
+      periodSeconds: 10
+      successThreshold: 1
+      timeoutSeconds: 1
+    resources:
+      limits:
+        cpu: "1"
+        memory: 1Gi
+      requests:
+        cpu: 50m
+        memory: 64Mi
```
Comment (antonincms) [inline charts/helm-controller-operator/values.yaml:108]:
```suggestion
    resources:
    #  limits:
    #    cpu: "1"
    #    memory: 1Gi
    #  requests:
    #    cpu: 50m
    #    memory: 64Mi
```

**Review by antonincms — CHANGES_REQUESTED** [review]

File: charts/helm-controller-operator/values.yaml (hunk 17 above)
Comment (antonincms) [reply charts/helm-controller-operator/values.yaml:164]: You did not resolve this issue yet

File: charts/helm-controller-operator/values.yaml [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+  # -- Override rtmp operator ServiceAccount name.
+  name: ""
+  # -- Annotations added to the ServiceMonitor.
+  annotations: {}
+
+service:
+  type: ClusterIP
+  port: 80
+  clusterIps:
+    - 10.96.163.142
```
Comment (antonincms) [inline charts/helm-controller-operator/values.yaml:164]: Service should be specific to either one. Furthermore clusterIps is not needed and type is not used.
```suggestion
```

File: charts/helm-controller-operator/templates/helmController/_helpers-helm-controller.tpl [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+app.kubernetes.io/component: helmController
+{{- end }}
+
+{{/*
+querier selector labels
+*/}}
+{{- define "helm-controller-operator.helmControllerSelectorLabels" -}}
+{{ include "helm-controller-operator.selectorLabels" . }}
+app.kubernetes.io/component: helmController
+{{- end }}
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/helmController/_helpers-helm-controller.tpl:22]:
```suggestion
{{/*
helm-controller fullname
*/}}
{{- define "helm-controller-operator.helmControllerFullname" -}}
{{ include "helm-controller-operator.fullname" . }}-helm
{{- end }}

{{/*
helm-controller common labels
*/}}
{{- define "helm-controller-operator.helmControllerLabels" -}}
{{ include "helm-controller-operator.labels" . }}
app.kubernetes.io/component: helm-controller
{{- end }}

{{/*
helm-controller selector labels
*/}}
{{- define "helm-controller-operator.helmControllerSelectorLabels" -}}
{{ include "helm-controller-operator.selectorLabels" . }}
app.kubernetes.io/component: helm-controller
{{- end }}
```

File: charts/helm-controller-operator/templates/sourceController/_helpers-source-controller.tpl [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+app.kubernetes.io/component: sourceController
+{{- end }}
+
+{{/*
+querier selector labels
+*/}}
+{{- define "helm-controller-operator.sourceControllerSelectorLabels" -}}
+{{ include "helm-controller-operator.selectorLabels" . }}
+app.kubernetes.io/component: sourceController
+{{- end }}
```
Comment (antonincms) [inline charts/helm-controller-operator/templates/sourceController/_helpers-source-controller.tpl:22]:
```suggestion
{{/*
source-controller fullname
*/}}
{{- define "helm-controller-operator.sourceControllerFullname" -}}
{{ include "helm-controller-operator.fullname" . }}-source
{{- end }}

{{/*
source-controller common labels
*/}}
{{- define "helm-controller-operator.sourceControllerLabels" -}}
{{ include "helm-controller-operator.labels" . }}
app.kubernetes.io/component: source-controller
{{- end }}

{{/*
source-controller selector labels
*/}}
{{- define "helm-controller-operator.sourceControllerSelectorLabels" -}}
{{ include "helm-controller-operator.selectorLabels" . }}
app.kubernetes.io/component: source-controller
{{- end }}
```

File: charts/helm-controller-operator/values.yaml [hunk 28]
```diff
@@ -0,0 +1,164 @@
+# Default values for helm-controller-manager
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+# -- ****** Value for the helm-controller *******
```
Comment (antonincms) [inline charts/helm-controller-operator/values.yaml:5]:
```suggestion
# ****** Value for the helm-controller *******
```

File: charts/helm-controller-operator/values.yaml [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+      seccompProfile:
+        type: RuntimeDefault
+
+    podSecurityContext:
+      fsGroup: 1337
+    podAnnotations: {}
+      # prometheus.io/scrape: "true"
+      # prometheus.io/port: "http-metrics"
+
+# -- ****** Value for the source-controller *******
```
Comment (antonincms) [inline charts/helm-controller-operator/values.yaml:62]: -- are made to document individual values

```suggestion
# Value for the source-controller
```

File: charts/helm-controller-stack/Chart.yaml [hunk 30]
```diff
@@ -0,0 +1,24 @@
...
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 0.1.0
+
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+# It is recommended to use it with quotes.
+appVersion: "1.16.0"
```
Comment (vincentmrg) [inline charts/helm-controller-stack/Chart.yaml:24]: Seems there is no app version
```suggestion
```

File: charts/helm-controller-stack/templates/sourceController/service.yaml [hunk 31]
```diff
@@ -0,0 +1,39 @@
...
+    - name: http-metrics
+      port: 8080
+      targetPort: http-metrics
+      protocol: TCP
+    {{- end }}
+  selector:
+    {{- include "helm-controller-stack.sourceControllerSelectorLabels" . | nindent 4 }}
+
+---
+
```
Comment (vincentmrg) [inline charts/helm-controller-stack/templates/sourceController/service.yaml:25]:
```suggestion
```

File: charts/helm-controller-stack/values.yaml [hunk 32]
```diff
@@ -0,0 +1,158 @@
...
+# Value for the helm-controller
+helmController:
+  manager:
+    # -- Number of desired pods.
+    replicaCount: 1
+    # -- Pods annotation.
+    podannotations: {}
+      # prometheus.io/port: "8080"
+      # prometheus.io/scrape: "true"
+    podAnnotations: {}
```
Comment (vincentmrg) [inline charts/helm-controller-stack/values.yaml:14]: duplicate, which one is used ?

File: charts/helm-controller-stack/README.md [hunk 33]
```diff
@@ -0,0 +1,103 @@
+# helm-controller-stack
+
+![Version: 0.1.0](https://img.shields.io/badge/Version-0.1.0-informational?style=flat-square) ![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square) ![AppVersion: 1.16.0](https://img.shields.io/badge/AppVersion-1.16.0-informational?style=flat-square)
+
+A Helm controller operator chart for Kubernetes
+
+Have a look on official documentation [Using Helm Operator in Kubernetes clusters](https://github.com/fluxcd/helm-controller)
```
Comment (vincentmrg) [inline charts/helm-controller-stack/README.md:7]: The description of this chart is quite minimal, a short paragraph explaining what it is, with a link to the flux repository might be relevant here.

File: charts/helm-controller-stack/Chart.yaml [hunk 34]
```diff
@@ -0,0 +1,18 @@
+apiVersion: v2
+name: helm-controller-stack
+description: A Helm controller operator chart for Kubernetes allowing one to declaratively manage Helm chart releases
```
Comment (vincentmrg) [inline charts/helm-controller-stack/Chart.yaml:3]:
```suggestion
description: A Helm controller chart for Kubernetes allowing one to declaratively manage Helm chart releases
```

File: charts/helm-controller-stack/templates/helmController/deployment.yaml [hunk 35]
```diff
@@ -0,0 +1,63 @@
...
+        {{- range $key, $value := .Values.helmController.manager.podAnnotations }}
+        {{ $key }}: {{ $value | quote }}
+        {{- end }}
+      labels:
+        {{- include "helm-controller-stack.helmControllerLabels" . | nindent 8 }}
+    spec:
+      containers:
+      - args:
+        - --watch-all-namespaces
+        - --enable-leader-election
```
Comment (antonincms) [inline charts/helm-controller-stack/templates/helmController/deployment.yaml:25]: You don't use the extraArgs present in values ?

File: charts/helm-controller-stack/templates/sourceController/deployment.yaml [hunk 36]
```diff
@@ -0,0 +1,70 @@
...
+        {{- end }}
+      labels:
+        {{- include "helm-controller-stack.sourceControllerLabels" . | nindent 8 }}
+    spec:
+      containers:
+      - args:
+        - --watch-all-namespaces
+        - --enable-leader-election
+        - --storage-path=/data
+        - --storage-adv-addr=source-controller.$(RUNTIME_NAMESPACE).svc.cluster.local.
```
Comment (antonincms) [inline charts/helm-controller-stack/templates/sourceController/deployment.yaml:27]: Idem, you should add an extraArgs

**Review by antonincms — APPROVED** [review]

**Pull request merged**